File: src/main/kotlin/com/wynnlab/spells/mage/MageMain.kt
package com.wynnlab.spells.mage

import com.wynnlab.spells.Spell
import com.wynnlab.spells.SpellData
import org.bukkit.entity.Player
import org.bukkit.Sound
import org.bukkit.entity.Mob
import com.wynnlab.util.LocationIterator
import org.bukkit.Particle

class MageMain(player: Player) : Spell(player, 1, SpellData.METEOR) {
    private val hit: MutableList<Mob> = mutableListOf()

    override fun tick() {
        val l1 = player.eyeLocation.clone().add(0.0, -0.5, 0.0)
        val l2 = player.eyeLocation.clone().add(player.eyeLocation.direction.clone().multiply(7)).clone().add(0.0, -0.5, 0.0)

        for (loc in LocationIterator(l1, l2, player.eyeLocation.direction.clone(), 0.5)) {
            player.spawnParticle(if (clone) Particle.SQUID_INK else Particle.CLOUD, loc, 2, 0.0, 0.0, 0.0, 0.1)
            player.spawnParticle(if (clone) Particle.SPELL_WITCH else Particle.CRIT, loc, 1, 0.0, 0.0, 0.0, 0.0)
            player.spawnParticle(Particle.CRIT_MAGIC, loc, 1, 0.0, 0.0, 0.0, 0.1)
            for (e in loc.getNearbyEntities(0.5, 0.5, 0.5)) {
                if (e is Player) {
                    continue
                }
                if (e is Mob) {
                    if (hit.contains(e)) {
                        continue
                    }
                    hit.add(e)
                    e.damage(2.0, player)
                }
            }
        }

        player.playSound(player.location, Sound.ENTITY_PLAYER_ATTACK_SWEEP, 1f, 1.5f)
        if (clone) player.playSound(player.location, Sound.ENTITY_SHULKER_SHOOT, 0.5f, 1.5f)
    }
}

File: src/main/resources/classes/mage/mage_main.py
from org.bukkit import Particle, Sound

from com.wynnlab.spells import PySpell
from com.wynnlab.util import LocationIterator


class Spell(PySpell):
    def __init__(self):
        self.hit = set()

    def tick(self):
        l1 = self.player.getEyeLocation().clone().add(0, -.5, 0)
        l2 = self.player.getEyeLocation().clone().add(self.player.getEyeLocation().getDirection().multiply(7)).add(0, -.5, 0)

        self.sound(Sound.ITEM_TRIDENT_THROW, 1, 1.5)
        self.sound(Sound.ITEM_TRIDENT_RIPTIDE_3, .2, 1)
        if self.clone:
            self.sound(Sound.ENTITY_SHULKER_SHOOT, .5, 1.5)

        for l in LocationIterator(l1, l2, self.player.getEyeLocation().getDirection(), .5):
            self.particle(l, Particle.SQUID_INK if self.clone else Particle.CLOUD, 2, 0, 0, 0, .1)
            self.particle(l, Particle.SPELL_WITCH if self.clone else Particle.CRIT, 1, 0, 0, 0, 0)
            self.particle(l, Particle.CRIT_MAGIC, 1, 0, 0, 0, .1)

            for e in self.nearbyMobs(l, .5, .5, .5):
                if e in self.hit:
                    continue
                self.hit.add(e)

                self.damage(e, True, 1)
                self.knockback(e, .5)